### DISCUSSION TITLE
data-pagefind-body set via frontmatter

### DISCUSSION THREAD

**Discussion by Karlstens** [ATQQ/sugar-blog#250] [Ideas]
Thanks for this plugin. I have questions regarding the way data-pagefind-body is handled, and how we can target specific pages via frontmatter.

Within the index.mjs / index.js files;

`node_modules\vitepress-plugin-pagefind\dist\index.mjs`

I note this snippet;
```
// 添加检索的内容标识 ( Add content identifiers for retrieval)
async transform(code, id) {
	if (id.endsWith("theme-default/Layout.vue")) {
		return code.replace("<VPContent>", "<VPContent data-pagefind-body>");
	}
	...
	return `${code}
}
```

This places data-pagefind-body in all VitePress pages during the build process allowing for each of these pages to be searched, as per PageFind docs; https://pagefind.app/docs/indexing/#limiting-what-sections-of-a-page-are-indexed

I'm thinking that it would be useful to index project pages based on VitePress frontmatter content. Perhaps for example, only index pages;

```
---
{
  "title": "My Amazing Webpage",
  "category": "About",
  "description": "About my amazing website and contact details",
  "data-pagefind-body": true
}
---
```

This would result in any page containing `  "data-pagefind-body": true` to be indexed, and all other pages ignored during the index build step. 

I figured I'd ask about this before getting to involved in editing the index.mjs file.

**Comment by ATQQ** [comment]
Do you want a frontmatter to control the page not to be indexed?

VitePress is built using SSG. Each generated page contains duplicate content regardless of NavBar, sideBar, etc., which will cause duplication in the generated index. When the keywords you search for are in sideBar or NavBar, the search results will appear. All pages will appear.

This VPContent component only contains the article content generated by markdown, so I added this tag to it to prevent duplicate parts on different pages from being generated and retrieved.

If you just want the article not to appear in search results, you can set

```md
---
publish: false
---
```

Or add a custom `forntmatter` value and then implement custom filtering rules through filter

**Reply by Karlstens** [reply]
Ahh excellent, `publish: false` is exactly what I needed. Thank you. 

I was initially confused as during the build, I noticed that all indexed values still remain the same, regardless of if pages are marked as `publish: false` or not.

```
[Building search indexes]   
Total:
  Indexed 1 language
  Indexed 13 pages
  Indexed 390 words
  Indexed 2 filters
  Indexed 1 sort
  ```
  
For example, as I updated more pages with `publish : false` I was expecting to see "Indexed 12 pages... Indexed 11 pages" as I save each markdown and test a re-build/serve.
  
However, upon checking via serve, the pages marked "publish : false" are not returning - so functionally that's working. I am interested in understanding why a page is still "indexed" yet doesn't return in the search when frontmatter contains `publish: false` (which is desired). 

Thank you again!

**Reply by ATQQ** [reply]
Yes, this is the case now. You can see the following code for how `publish` works.

https://github.com/ATQQ/sugar-blog/blob/1ab7987aa71b5870c796eb612827802fb38a0886/packages/vitepress-plugin-pagefind/src/Search.vue#L153-L155

Its function is only to filter out the search content, and the command to generate the index will not be affected. I will study this to see if it can be done without generating the index to achieve the desired effect.

The actual meaning of `publish` is to indicate whether the article is published. False means that the creation has not been completed, but it can still participate in the construction and VitePress deployment, but does not allow users to access it.

This is equivalent to not displaying uncreated articles in the search results.

**Comment by ATQQ** [comment]
I read the documentation and there should be at least 3 ways to achieve it. I will study it to see which one is more suitable for VitePress, which is simpler to implement and has better compatibility.

* CLI Glob option
* Node JS API
* data-pagefind-body attr

**Reply by Karlstens** [reply]
I'm all ready to test this with you too, let me know and I'll make time to help.

**Edit:**
In doing a little searching on frontmatter publish boolean, there's talk mentioning that it's desired to prevent VitePress from publishing that markdown file - in my testing, VitePress publishes the page regardless of frontmatter publish boolean. 

However, PageFind and Vitepress-Plugin-Pagefind act upon frontmatter publish boolean set to false preventing that page from returning via search (which is the functionality that I'm after). 

A thought - perhaps it could be considered to add frontmatter "pagefind-indexed : true/false", just in case there's a conflict with "publish : true/false".

**Reply by ATQQ** [reply]
try `vitepress-plugin-pagefind@0.4.4`

### Example 5: exclude indexing some pages
set frontmatter `pagefind-indexed: false` to exclude.
```md
---
pagefind-indexed: false
---
```

---
https://github.com/ATQQ/sugar-blog/pull/255

**Reply by Karlstens** [reply]
Initial tests are looking good! 

With this update implemented, I love how the pagefind build report adjusts the page count accordingly. Now we can actually see the count of pages being indexed reducing when employing `pagefind-indexed: false`.

Thank you, this is most excellent. 🥳
